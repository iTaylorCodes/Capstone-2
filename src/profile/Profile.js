import { useState, useContext } from "react";
import Alert from "../alert/Alert";
import { NeighborhoodApi } from "../api/api";
import UserContext from "../auth/UserContext";
import { useDispatch } from "react-redux";

/** Profile.
 *
 * Shows a form displaying user profile details.
 * Handles changing user details.
 * On submission, updates user details.
 *
 * Also contains Delete Account button.
 * On click, deletes the current user's account from database.
 *
 * Routed at /profile.
 */
const Profile = ({ logout }) => {
  const dispatch = useDispatch();
  const { currentUser, setCurrentUser } = useContext(UserContext);
  const [formData, setFormData] = useState({
    username: currentUser.username,
    firstName: currentUser.firstName,
    lastName: currentUser.lastName,
    email: currentUser.email,
    password: "",
  });
  const [formErrors, setFormErrors] = useState([]);

  const [isSaved, setIsSaved] = useState(false);

  // Handles form submit
  const handleSubmit = async (e) => {
    e.preventDefault();

    let profileData = {
      firstName: formData.firstName,
      lastName: formData.lastName,
      email: formData.email,
      password: formData.password,
    };

    let username = formData.username;
    let updatedUser;

    try {
      updatedUser = await NeighborhoodApi.updateProfile(username, profileData);
    } catch (errors) {
      setFormErrors(errors);
      return;
    }

    setFormData((data) => ({ ...data, password: "" }));
    setFormErrors([]);
    setIsSaved(true);

    // Reloads user data across site
    setCurrentUser(updatedUser);
  };

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((data) => ({
      ...data,
      [name]: value,
    }));
    setFormErrors([]);
    setIsSaved(false);
  };

  const deleteAcc = async (e) => {
    e.preventDefault();

    try {
      await NeighborhoodApi.deleteProfile(currentUser.username);
    } catch (errors) {
      setFormErrors(errors);
      return;
    }

    dispatch({
      type: "ACCOUNT_DELETED",
    });
    logout();
  };

  return (
    <div className="row justify-content-center mb-3">
      <div className="container col-10 col-sm-9 col-md-8 col-lg-7">
        <div className="row mt-3 ms-1 me-1">
          <div className="col-5">
            <h1>Profile</h1>
          </div>
          <div className="d-flex justify-content-end col-7">
            <button
              className="btn btn-outline-danger btn-sm"
              style={{ height: "35px" }}
              onClick={deleteAcc}
            >
              Delete Account
            </button>
          </div>
        </div>
        <div className="card">
          <form onSubmit={handleSubmit}>
            <fieldset disabled className="m-3">
              <label htmlFor="username" className="form-label">
                Username
              </label>
              <input
                id="username"
                name="username"
                value={formData.username}
                onChange={handleChange}
                className="form-control"
                placeholder="Username"
                autoComplete="username"
              />
            </fieldset>
            <div className="m-3">
              <label htmlFor="first-name" className="form-label">
                First Name
              </label>
              <input
                id="first-name"
                name="firstName"
                value={formData.firstName}
                onChange={handleChange}
                className="form-control"
                placeholder="First Name"
                autoComplete="firstName"
              />
            </div>
            <div className="m-3">
              <label htmlFor="last-name" className="form-label">
                Last Name
              </label>
              <input
                id="last-name"
                name="lastName"
                value={formData.lastName}
                onChange={handleChange}
                className="form-control"
                placeholder="Last Name"
                autoComplete="lastName"
              />
            </div>
            <div className="m-3">
              <label htmlFor="email" className="form-label">
                Email
              </label>
              <input
                id="email"
                name="email"
                value={formData.email}
                onChange={handleChange}
                className="form-control"
                placeholder="Email"
                autoComplete="email"
              />
            </div>
            <div className="m-3">
              <label htmlFor="password" className="form-label fw-bold">
                Confirm password to save changes:
              </label>
              <input
                id="password"
                name="password"
                type="password"
                value={formData.password}
                onChange={handleChange}
                className="form-control"
                placeholder="Password"
                autoComplete="password"
              />
            </div>

            {formErrors.length ? (
              <Alert type="danger" messages={formErrors} />
            ) : null}

            {isSaved ? (
              <Alert type="success" messages={["Updated successfully!"]} />
            ) : null}

            <div className="m-3 d-flex justify-content-end">
              <button className="btn btn-primary btn-lg" type="submit">
                Save
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};

export default Profile;
