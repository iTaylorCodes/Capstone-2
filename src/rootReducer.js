import { v4 as uuidv4 } from "uuid";

const INITIAL_STATE = {
  accountWasDeleted: false,
  recentSearches: [],
};

const rootReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case "ACCOUNT_DELETED":
      return {
        ...state,
        accountWasDeleted: true,
      };

    case "RESET_ACCOUNT_DELETED":
      return {
        ...state,
        accountWasDeleted: false,
      };

    case "NEW_SEARCH":
      if (state.recentSearches.length >= 5) {
        state.recentSearches.shift();
      }

      return {
        ...state,
        recentSearches: [
          ...state.recentSearches,
          {
            city: action.searchTerm,
            image: action.image,
            scores: action.scores,
            id: uuidv4(),
          },
        ],
      };

    default:
      return state;
  }
};

export default rootReducer;
