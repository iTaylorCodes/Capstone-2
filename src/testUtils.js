import UserContext from "./auth/UserContext";

const testUser = {
  username: "testuser",
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
  password: "password",
};

const UserProvider = ({ children, currentUser = testUser }) => (
  <UserContext.Provider value={{ currentUser }}>
    {children}
  </UserContext.Provider>
);

export { UserProvider };
