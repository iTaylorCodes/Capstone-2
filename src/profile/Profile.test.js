import { render, act, fireEvent } from "@testing-library/react";
import Profile from "./Profile";
import { UserProvider } from "../testUtils";
import * as redux from "react-redux";
import { Provider } from "react-redux";
import { createStore } from "redux";
import rootReducer from "../rootReducer";

const store = createStore(rootReducer);

const spy = jest.spyOn(redux, "useSelector");
spy.mockReturnValue({ accountWasDeleted: false });

it("renders without crashing", () => {
  render(
    <Provider store={store}>
      <UserProvider>
        <Profile />
      </UserProvider>
    </Provider>
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <Provider store={store}>
      <UserProvider>
        <Profile />
      </UserProvider>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("shows bootstrap style alerts for errors", async () => {
  await act(async () => {
    const { getByText, getByLabelText } = render(
      <Provider store={store}>
        <UserProvider>
          <Profile />
        </UserProvider>
      </Provider>
    );

    const emailInput = getByLabelText("Email");
    await fireEvent.change(emailInput, { target: { value: "test@test.com" } });

    const button = getByText("Save");
    await fireEvent.click(button);

    expect(getByLabelText("Email").value).toBe("test@test.com");
  });
});

it("shows bootstrap style alerts on success", async () => {
  await act(async () => {
    const { getByText, getByLabelText } = render(
      <Provider store={store}>
        <UserProvider>
          <Profile />
        </UserProvider>
      </Provider>
    );

    const emailInput = getByLabelText("Email");
    await fireEvent.change(emailInput, {
      target: { value: "tester@test.net" },
    });

    const passwordInput = getByLabelText("Confirm password to save changes:");
    await fireEvent.change(passwordInput, { target: { value: "password" } });

    const button = getByText("Save");
    await fireEvent.click(button);

    expect(getByLabelText("Email").value).toBe("tester@test.net");
  });
});

it("can delete user accounts", async () => {
  await act(async () => {
    const { getByText } = render(
      <Provider store={store}>
        <UserProvider>
          <Profile />
        </UserProvider>
      </Provider>
    );

    const button = getByText("Delete Account");
    await fireEvent.click(button);

    spy.mockReturnValue({ accountWasDeleted: true });

    expect(spy).toBeTruthy();
  });
});
