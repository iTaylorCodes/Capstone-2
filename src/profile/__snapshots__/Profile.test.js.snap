// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="row justify-content-center mb-3"
  >
    <div
      class="container col-10 col-sm-9 col-md-8 col-lg-7"
    >
      <div
        class="row mt-3 ms-1 me-1"
      >
        <div
          class="col-5"
        >
          <h1>
            Profile
          </h1>
        </div>
        <div
          class="d-flex justify-content-end col-7"
        >
          <button
            class="btn btn-outline-danger btn-sm"
            style="height: 35px;"
          >
            Delete Account
          </button>
        </div>
      </div>
      <div
        class="card"
      >
        <form>
          <fieldset
            class="m-3"
            disabled=""
          >
            <label
              class="form-label"
              for="username"
            >
              Username
            </label>
            <input
              autocomplete="username"
              class="form-control"
              id="username"
              name="username"
              placeholder="Username"
              value="testuser"
            />
          </fieldset>
          <div
            class="m-3"
          >
            <label
              class="form-label"
              for="first-name"
            >
              First Name
            </label>
            <input
              autocomplete="firstName"
              class="form-control"
              id="first-name"
              name="firstName"
              placeholder="First Name"
              value=""
            />
          </div>
          <div
            class="m-3"
          >
            <label
              class="form-label"
              for="last-name"
            >
              Last Name
            </label>
            <input
              autocomplete="lastName"
              class="form-control"
              id="last-name"
              name="lastName"
              placeholder="Last Name"
              value=""
            />
          </div>
          <div
            class="m-3"
          >
            <label
              class="form-label"
              for="email"
            >
              Email
            </label>
            <input
              autocomplete="email"
              class="form-control"
              id="email"
              name="email"
              placeholder="Email"
              value="test@test.net"
            />
          </div>
          <div
            class="m-3"
          >
            <label
              class="form-label fw-bold"
              for="password"
            >
              Confirm password to save changes:
            </label>
            <input
              autocomplete="password"
              class="form-control"
              id="password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="m-3 d-flex justify-content-end"
          >
            <button
              class="btn btn-primary btn-lg"
              type="submit"
            >
              Save
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
